File: antia_numerical_methods/c/fun/betser.c
/*	To calculate the incomplete Beta function I_x(a,b) using
	the infinite series, called by BETAP

	a,b : (input) Arguments for the complete Beta function
	x : (input) Upper limit of integration defining the incomplete
			Beta function

	Required routines : gammaln
*/


#include <math.h>
#include <stdio.h>

double gammaln(double x);

double betser(double a, double b, double x)
{
	double t,t1,tmax,b1,s3;
	int i;

	s3=1; t=a; tmax=a; i=-1;
	while(i<500 && fabs(t/(a+i+1))>1.e-15)
	{
		i=i+2; t1=t*x*(i-b)/i; t=t1*x*(i+1-b)/(i+1);
		s3=s3+t1/(a+i)+t/(a+i+1);
		if(fabs(t/(a+i+1))>tmax) tmax=fabs(t/(a+i+1));
	}
	if(s3<1.e-16*tmax) printf("  Roundoff error while evaluating the infinite series for incomplete Beta function at\n a= %e  b= %e  x= %e sum= %e  Max term= %e\n",a,b,x,s3,tmax);
	b1=a*log(x)+log(s3)-log(a);
	b1=b1+gammaln(a+b)-gammaln(a)-gammaln(b);
	return exp(b1);
}

File: projects/antia_numerical_methods@Beliavsky/f77/sub16/betser.f
!	To calculate the incomplete Beta function I_x(a,b) using
!	the infinite series, called by BETAP
!
!	A,B : (input) Arguments for the complete Beta function
!	X : (input) Upper limit of integration defining the incomplete
!                       Beta function
!
!	Required routines : GAMMAL

      FUNCTION BETSER(A,B,X)
      IMPLICIT REAL*16(A-H,O-Z)

51    FORMAT(' ** Roundoff error while evaluating the infinite series',
     1  ' for incomplete Beta function at',/,'  A=',1PD12.4,'  B=',
     2  D12.4,'  X=',D12.4,'  SUM=',D12.4,'  MAX TERM=',D12.4)

      S3=1
      T=A
      TMAX=A
      DO I=1,500,2
        T1=T*X*(I-B)/I
        T=T1*X*(I+1-B)/(I+1)
        S3=S3+(T1/(A+I)+T/(A+I+1))
        IF(ABS(T/(A+I+1)).GT.TMAX) TMAX=ABS(T/(A+I+1))
        IF(ABS(T/(A+I+1)).LT.1.Q-15) EXIT
      ENDDO
      IF((S3).LT.1.Q-16*TMAX) PRINT 51,A,B,X,S3,TMAX
      B1=A*LOG(X)+LOG(S3)-LOG(A)
      B1=B1+GAMMAL(A+B)-GAMMAL(A)-GAMMAL(B)
      BETSER=EXP(B1)
      END
